Translated techlab 3.5 to english (#31)
appuio/techlab: labs/09_dockerbuild_webhook.md

> # Lab 9: Directly integrate code changes via Webhook

Reviewer: ```suggestion
# Lab 9: Integrating Code Changes Via Webhook
```

> In this lab, we'll show you the Docker Build Workflow by example, and you'll learn how to push a build and deploy the application to OpenShift with a push into the Git repository.

Reviewer: ```suggestion
In this lab we'll show you the Docker build workflow and how to push a build and deploy the application to OpenShift with a push into a Git repository.
```

> ## Task: LAB9.1: Preparation of Github Account and Fork

Reviewer: ```suggestion
## Task: LAB9.1: GitHub Preparations
```

### Github Account

In order to make changes to the source code of our example application, you need a separate GitHub account. If you do not already have one, set up an account at https://github.com/.

> ### Fork Example-Project

Reviewer: ```suggestion
### Forking the Example Project
```

> **Example-Project:** https://github.com/appuio/example-php-docker-helloworld

Reviewer: ```suggestion
**Example project:** https://github.com/appuio/example-php-docker-helloworld
```

> Go to [GitHub Projekt-Seite](https://github.com/appuio/example-php-docker-helloworld) and [fork](https://help.github.com/articles/fork-a-repo/) the project.

Reviewer: ```suggestion
Go to [the GitHub project page](https://github.com/appuio/example-php-docker-helloworld) and [fork](https://help.github.com/articles/fork-a-repo/) the project.
```

![Fork](../images/lab_9_fork_example.png)

> You now have

Reviewer: ```suggestion
You now have a fork of the example project at
```

```txt
https://github.com/[YourGitHubUser]/example-php-docker-helloworld
```

> a fork of the Example project that you can expand as you want.

Reviewer: ```suggestion
which you can modify as you want.
```

> ## Deployen your own fork

Reviewer: ```suggestion
## Deploying the Forked Application
```

Create a new project:

```bash
oc new-project [USER]-example4
```

Create a new app for your fork.  **Note:** Replace `[YourGithubUser]` with the name of your GitHub account:

```bash
oc new-app https://github.com/[YourGithubUser]/example-php-docker-helloworld.git --strategy=docker --name=appuio-php-docker-ex
```

By means of the parameter  `--strategy=docker` we explicitly tell the `oc new-app` command to look for a Dockerfile in the specified Git repository and use it for the build.

Now expose the service with:

```bash
oc expose service appuio-php-docker-ex
```

> ## Task: LAB9.2: Set up web hook on GitHub

Reviewer: ```suggestion
## Task: LAB9.2: Set Up the Web Hook
```

> When creating the app, BuildConfig (bc) directly defined webhooks. You can do this by using the following command:

Reviewer: ```suggestion
During the creation of the app, webhooks have automatically been created and inserted into the BuildConfig (bc). Let's get the webhook URLs using the following command:
```

```

> You can also copy the GitHub WebHook from the Web Console. To do this, go to the appropriate build using Builds → Builds and select the Configuration tab:

Reviewer: ```suggestion
You can also copy the GitHub webhook from the web console. To do this, go to the appropriate build using Builds → Builds and select the Configuration tab:
```

![Webhook](../images/lab_9_webhook_ose3.png)

> Copy the GitHub [Webhook](https://developer.github.com/webhooks/) URL and paste it into GitHub.

Reviewer: ```suggestion
Copy the GitHub [webhook](https://developer.github.com/webhooks/) URL and paste it into GitHub.
```

In your project, click Settings:
![Github Webhook](../images/lab_09_webhook_github1.png)

Click Webhooks & services:
![Github Webhook](../images/lab_09_webhook_github2.png)

Add a WebHook:
![Github Webhook](../images/lab_09_webhook_github3.png)

> Insert the appropriate GitHub WebHook URL from your OpenShift project and "disable" the SSL verification. On the Lab platform we have only self-signed certificates.

Reviewer: ```suggestion
Insert the appropriate GitHub Webhook URL from your OpenShift project.
```

![Github Webhook](../images/lab_09_webhook_github4.png)

> From now, all pushes on your GitHub repository triggers a build and then deploy the code changes directly to the OpenShift platform.

Reviewer: ```suggestion
From now on all pushes to your GitHub repository triggers a build and then deploys the code changes directly to the OpenShift platform.
```

> ## Task: LAB9.3: Adjust the code

Reviewer: ```suggestion
## Task: LAB9.3: Code Update
```

    </div>
```

Now push your changes:

```bash
$ git add .
$ git commit -m "updated Hello"
$ git push
[...]
```

As an alternative you can edit the file directly on GitHub:
![Github Webhook](../images/lab_9_edit_on_github.png)

As soon as you pushed your changes, OpenShift will start a new build of the source code:

```bash
oc get builds
```

and deployes it afterwards.

## Task: LAB9.4: Rollback

With OpenShift, different software versions can be activated and deactivated by simply starting another version of the image.

The command `oc rollback` and `oc rollout` are used for this purpose.

To run a rollback, you need the name of the DeploymentConfig:

```bash
$ oc get dc

NAME                   REVISION   DESIRED   CURRENT   TRIGGERED BY
appuio-php-docker-ex   4          1         1         config,image(appuio-php-docker-ex:latest)

```

Use the following command to roll back to the previous version:

```bash
$ oc rollback appuio-php-docker-ex
#3 rolled back to appuio-php-docker-ex-1
Warning: the following images triggers were disabled: appuio-php-docker-ex:latest
  You can re-enable them with: oc set triggers dc/appuio-php-docker-ex --auto
```

> As soon as the deployment of the old version is done, you can check in your Browser if the original Titel **Hello APPUiO** is back.

Reviewer: ```suggestion
As soon as the deployment of the old version is done, you can check in your browser that the original title **Hello APPUiO** is back.
```

**Hint:** The automatic deployments of new versions are now switched off for this application to prevent unintentional changes after the rollback. To turn automatic deployment back on, run the following command:

```bash
oc set triggers dc/appuio-php-docker-ex --auto
```

---

**Ende Lab 9**

> <p width="100px" align="right"><a href="10_persistent_storage.md">Persistent Storage and use for database →</a></p>

Reviewer: ```suggestion
<p width="100px" align="right"><a href="10_persistent_storage.md">Persistent Storage →</a></p>
```
